src/css3.rs

```diff
@@ -1,120 +1,959 @@
-//! string_1     \"([^\n\r\f\\"]|\\{nl} | {escape})*\"
-//! string_2     \'([^\n\r\f\\']|\\{nl} | {escape})*\'
-//! invalid     {invalid_1} | {invalid_2}
-//! invalid_1    \"([^\n\r\f\\"] | \\{nl} | {escape})*
-//! invalid_2    \'([^\n\r\f\\'] | \\{nl} | {escape})*
-//! nl          \n |\r\n |\r |\f
-//! w           [ \t\r\n\f]*
-//! ```
-//!
-//! # Backus Naur Form (BNF)
-//!
-//! ```txt
-//! StyleSheet
-//!     : [CDO | CDC | Statement]
-//!     ;
-//!
-//! Statement
-//!     : RuleSet
-//!     | AtRule
-//!     ;
-//!
-//! AtRule
-//!     : ATKEYWORD Any* [Block | SEMICOLON]
-//!     ;
-//!
-//! Block
-//!     : LCURLY [Any | Block | ATKEYWORD | SEMICOLON]* RCURLY
-//!     ;
-//!
-//! RuleSet
-//!     : Selector? LCURLY Declaration? [SEMICOLON Declaration?]* RCURLY
-//!     ;
-//!
-//! Selector
-//!     : Any+
-//!     ;
-//!
-//! Declaration
-//!     : DELIM? Property COLON Value
-//!     ;
-//!
-//! Property
-//!     : IDENTY
-//!     ;
-//!
-//! Value
-//!     : [Any | Block | ATKEYWORD]+
-//!     ;
-//!
-//! Any
-//! :   [   IDENT
-//!     |   NUMBER
-//!     |   PERCENTAGE
-//!     |   DIMENSION
-//!     |   STRING
-//!     |   DELIM
-//!     |   URI
-//!     |   HASH
-//!     |   UNICODE-RANGE
-//!     |   INCLUDES
-//!     |   DASHMATCH
-//!     |   FUNCTION Any* RPAREN
-//!     |   LPAREN Any* RPAREN
-//!     |   LBRACKET Any* RBRACKET
-//!     ]
-//! ;
-//! ```
-//!
-pub mod new_parser;
-pub mod new_tokenizer;
-pub mod node;
-pub mod parser;
-pub mod tokenizer;
-pub mod tokens;
-pub mod unicode;
+mod new_tokenizer;
+mod unicode;
+mod nom_selector;
+mod ast;
+
+use crate::css3::new_tokenizer::{Token, TokenStreamer};
+use crate::css3::Error::{Syntax, UnexpectedEof};
+use core::fmt::Debug;
+use std::collections::HashMap;
+use log::{debug, trace};
+use thiserror::Error;
+use crate::css3::ast::Node;
+
+#[derive(Error, Debug)]
+pub enum Error {
+    #[error("syntax error: {0}")]
+    Syntax(String),
+    #[error("unexpected end of stream")]
+    UnexpectedEof,
+}
+
+// =================================================================================================
+// Tokenstream is a simple already-tokenized stream of tokens. This is used for example when
+// parsing a declaration, which is a list of tokens that are already known.
+struct TokenStream {
+    tokens: Vec<Token>,
+    index: usize,
+}
+
+impl TokenStream {
+    fn new(tokens: Vec<Token>) -> TokenStream {
+        TokenStream { tokens, index: 0 }
+    }
+
+    fn new_from_componentvalues(values: Vec<ComponentValue>) -> TokenStream {
+        let tokens = values.iter().filter(|v| {
+            if let ComponentValue::PreservedToken(_) = v {
+                true
+            } else {
+                false
+            }
+        }).map(|v| {
+            if let ComponentValue::PreservedToken(token) = v {
+                token.clone()
+            } else {
+                panic!("we should not be here");
+            }
+        }).collect::<Vec<Token>>().into();
```

Comment:
I think, you can remove the `into` here
```suggestion
        }).collect::<Vec<Token>>();
```